v3/src/components/graph/models/graph-controller.ts

```diff
@@ -23,51 +31,107 @@ export interface IGraphControllerProps {
   instanceId: string
   dotsRef: React.RefObject<SVGSVGElement>
+  v2Document?: CodapV2Document
 }
 
 export class GraphController {
   graphModel: IGraphModel
   layout: GraphLayout
   dataset: IDataSet | undefined
-  dataConfig: IDataConfigurationModel
   enableAnimation: React.MutableRefObject<boolean>
   instanceId: string
   dotsRef: React.RefObject<SVGSVGElement>
+  v2Document?: CodapV2Document
 
 
   constructor(props: IGraphControllerProps) {
     this.graphModel = props.graphModel
     this.layout = props.layout
     this.dataset = props.dataset
-    this.dataConfig = DataConfigurationModel.create()
     this.instanceId = props.instanceId
     this.enableAnimation = props.enableAnimation
     this.dotsRef = props.dotsRef
+    this.v2Document = props.v2Document
     if (this.dataset) {
-      this.dataConfig.setDataset(this.dataset)
+      this.graphModel.config.setDataset(this.dataset)
     }
     // Presumably a new dataset is now being used. So we have to set things up for an empty graph
     this.initializeGraph()
   }
 
   initializeGraph() {
-    const {graphModel, dataConfig, layout, dotsRef, enableAnimation, instanceId} = this
-    graphModel.setGraphProperties({
-      axes: {bottom: EmptyAxisModel.create({place: 'bottom'}), left: EmptyAxisModel.create({place: 'left'})},
-      plotType: 'casePlot', config: this.dataConfig
-    })
+    const {graphModel, layout, dotsRef, enableAnimation, instanceId, v2Document} = this,
+      dataConfig = graphModel.config
+    if (v2Document) {
+      this.processV2Document()
+    } else {
+      graphModel.setGraphProperties({
+        axes: {bottom: EmptyAxisModel.create({place: 'bottom'}),
+          left: EmptyAxisModel.create({place: 'left'})}, plotType: 'casePlot'
+      })
+      layout.setAxisScale('bottom', scaleOrdinal())
+      layout.setAxisScale('left', scaleOrdinal())
+    }
     if (dotsRef.current) {
       matchCirclesToData({
         caseIDs: dataConfig.cases, dotsElement: dotsRef.current,
         pointRadius: graphModel.getPointRadius(), enableAnimation, instanceId
       })
     }
-    layout.setAxisScale('bottom', scaleOrdinal())
-    layout.setAxisScale('left', scaleOrdinal())
+  }
+
+  processV2Document() {
+    const {graphModel, layout, /*dotsRef, enableAnimation,*/ dataset, v2Document} = this,
+      dataConfig = graphModel.config,
+      firstV2GraphComponent = v2Document?.components.find(aComp => aComp.type === 'DG.GraphView'),
+      storage = firstV2GraphComponent?.componentStorage as ICodapV2GraphStorage,
+      links = storage?._links_ || {},
+      attrTypes: Record<string, string> = {x: 'empty', y: 'empty', legend: 'empty'}
+    Object.keys(links).forEach(aKey => {
+      if (['xAttr', 'yAttr', 'legendAttr'].includes(aKey)) {
+        const match = aKey.match(/[a-z]+/),
+          attrPlace = (match ? match[0] : 'x') as GraphAttrPlace,
+          attrV2ID = (links[aKey] as unknown as IGuidLink<"DG.Attribute">).id,
```

Comment:
```suggestion
    Object.keys(links).forEach((aKey: keyof typeof links) => {
      if (['xAttr', 'yAttr', 'legendAttr'].includes(aKey)) {
        const match = aKey.match(/[a-z]+/),
          attrPlace = (match?.[0] ?? 'x') as GraphAttrPlace,
          attrV2ID = (links[aKey] as IGuidLink<"DG.Attribute">).id,
```
By declaring the type of the loop variable initially, you can reduce the need to cast down below. See https://imfaber.me/typescript-how-to-iterate-over-object-properties/ for details on the type.